test/e2e/operator.go

```diff
@@ -272,3 +277,102 @@ var _ = Describe("ARO Operator - Conditions", func() {
 		Expect(err).NotTo(HaveOccurred())
 	})
 })
+
+var _ = Describe("ARO Operator - Azure NSG Reconciler", func() {
+	var location, resourceGroup, subnet, vnet string
+	var correctNSG *string
+	var dummynsg mgmtnetwork.SecurityGroup
+	ctx := context.Background()
+	type subnetDescriptor struct {
+		resourceGroup string
+		vnetName      string
+		subnetName    string
+		location      string
+	}
+	enableReconcileNSG := func() {
+		instance, err := clients.AROClusters.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+		if !instance.Spec.Features.ReconcileNSGs {
+			instance.Spec.Features.ReconcileNSGs = true
+			_, err = clients.AROClusters.AroV1alpha1().Clusters().Update(ctx, instance, metav1.UpdateOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		}
+	}
+	getDescriptorFromProviderSpec := func(providerSpec *runtime.RawExtension) (*string, *subnetDescriptor, error) {
+		var spec azureproviderv1beta1.AzureMachineProviderSpec
+		err := json.Unmarshal(providerSpec.Raw, &spec)
+		if err != nil {
+			return nil, nil, err
+		}
+		return &spec.ResourceGroup, &subnetDescriptor{
+			resourceGroup: spec.NetworkResourceGroup,
+			vnetName:      spec.Vnet,
+			subnetName:    spec.Subnet,
+			location:      spec.Location,
+		}, nil
+	}
+	gatherNetworkInfo := func() {
+		machines, err := clients.MachineAPI.MachineV1beta1().Machines("openshift-machine-api").List(ctx, metav1.ListOptions{LabelSelector: "machine.openshift.io/cluster-api-machine-role=worker"})
+		Expect(err).NotTo(HaveOccurred())
+		subnetMap := make(map[subnetDescriptor]bool)
+		for _, machine := range machines.Items {
+			_, subnetDesc, err := getDescriptorFromProviderSpec(machine.Spec.ProviderSpec.Value)
+			if err != nil {
+				log.Fatal(err)
+			}
+			subnetMap[*subnetDesc] = true
+		}
+		for ws := range subnetMap {
+			location = ws.location
+			resourceGroup = ws.resourceGroup
+			subnet = ws.subnetName
+			vnet = ws.vnetName
+		}
+	}
```

Comment:
Do we actually need all this? My understanding of what this code is doing:

1. It lists all worker machine sets
2. It gathers unique subnets (`subnetMap`)
3. Then it take only one subnet from `subnetMap` and the further tests work with it. 

Why do we gather unique subnets if we need only one?

Reply:
I think we don't need to read machinesets here at all. Can we do something like this instead?

1. Read cluster object to get vnet id.
2. Get vnet object which contains all subnets
3. Modify one of the subnets.

See [`[Admin API] List Azure resources action`](https://github.com/Azure/ARO-RP/blob/e85e4a07adef15378b350f913800ae90cb2af018/test/e2e/adminapi_resources.go) for a refrence.

I have a feeling that it will significantly simplify the test.

Reply:
I have removed this code that relied on logic using in NSG controller and instead now use method you suggested.